# Transcriptome Analysis with kallisto

#### Autor: Luiz Carlos Vieira
#### Date: 25/03/2022


Kallisto rely on a pseudoalignment which does not identify the positions of the reads in the transcripts, only their potential 
transcripts of origin. It avoids doing an alignment of each read to a reference genome, using just the transcriptome sequences
as reference genome [1].

Standard methods for quantifying expression relied on mapping sequenced reads to a reference genome, assigning reads to a position
in the genome and the gene or isoform expression values are generated by counting the number of overlapping reads (overlapping the
features of interest) [2].


## Downloading Rna-Seq datasets
```bash
# GSE120561 sample experiment id
# Overall design: RNA-seq to measure gene expression in queen,
# queen right worker and actively laying worker ovaries
# tissue: ovary
# Library strategy:   RNA-Seq
# Library source: transcriptomic
# Library selection:  cDNA
# Instrument model:   Illumina HiSeq 2500

curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR790/006/SRR7908186/SRR7908186.fastq.gz -o SRR7908186_W1_Worker_Pool_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR790/007/SRR7908187/SRR7908187.fastq.gz -o SRR7908187_W2_Worker_Pool_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR790/008/SRR7908188/SRR7908188.fastq.gz -o SRR7908188_AW1_Active_Pool_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR790/009/SRR7908189/SRR7908189.fastq.gz -o SRR7908189_AW2_Active_Pool_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR790/001/SRR7908191/SRR7908191.fastq.gz -o SRR7908191_Q2_Queen_Pool_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR790/000/SRR7908190/SRR7908190.fastq.gz -o SRR7908190_Q1_Queen_Pool_1.fastq.gz

```


## Quality Control of reads
```bash
sample="GSE120561"; for i in $sample; do fastqc -t 4 ${i}/rawFastq/*.fastq -o /mnt/c/Users/luiz_/Downloads/transcriptomeAnalysis/${i}/QC; done

sample="GSE120561"; for i in $sample; do multiqc ${i}/QC/ -o ${i}/QC/
```


## Adding lncRNA into the transcriptome fasta
```bash
cat ref_genome/Amel_HAv3.1_rna.fna kallisto/lncRNA.fasta > kallisto/all_transcrits.fasta
```


## Building up a index of transcripts FASTA sequences (kallisto index)
```bash
kallisto index -i kallisto/index_trans kallisto/all_transcrits.fasta
```


## Verifying lncRNA among transcripts
```bash
makeblastdb -in kallisto/all_transcripts.fasta -dbtype nucl -input_type fasta -title "all_transcripts_Amel_HAv3.1" -out blast/db/all_transcripts_Amel_HAv3.1

blastn -db kallisto/all_transcripts_Amel_HAv3.1 -query kallisto/lncRNA.fasta -out kallisto/res_blast_lncRNA_x_all_trascripts.txt
```

Reference: Zheng Zhang, Scott Schwartz, Lukas Wagner, and Webb  
Miller (2000), "A greedy algorithm for aligning DNA sequences", J  
Comput Biol 2000; 7(1-2):203-14.  



Database: all_transcripts_Amel_HAv3.1  
           27,888 sequences; 98,646,852 total letters  



Query= lncRNA Clone_01-F  

Length=1152  
                                                                      Score        E  
Sequences producing significant alignments:                          (Bits)     Value  

lncRNA Clone_01-F                                                     2128       0.0  


>lncRNA Clone_01-F  
Length=1152  

 Score = 2128 bits (1152),  Expect = 0.0  
 Identities = 1152/1152 (100%), Gaps = 0/1152 (0%)  
 Strand=Plus/Plus  


## Verifying strandness of reads
```bash
SAMPLE=GSE
KALLISTO_INDEX=/mnt/c/Users/luiz_/Downloads/transcriptomeAnalysis/index/index_trans
OUT_DIR=/mnt/c/Users/luiz_/Downloads/transcriptomeAnalysis/${SAMPLE}/kallisto
SEQ_DIR=/mnt/c/Users/luiz_/Downloads/transcriptomeAnalysis/${SAMPLE}/rawFastq

# Make a subset from a fastqc
zcat ${SEQ_DIR}/SRR7770429_W01_mRNA_including_lncRNA_RNA-Seq_1.fastq.gz | head -n16000 > ${SEQ_DIR}/test_1.fastq
zcat ${SEQ_DIR}/SRR7770429_W01_mRNA_including_lncRNA_RNA-Seq_2.fastq.gz | head -n16000 > ${SEQ_DIR}/test_2.fastq

# Running kallisto quant
kallisto quant -i ${KALLISTO_INDEX} -o ${OUT_DIR}/test_un ${SEQ_DIR}/test_1.fastq ${SEQ_DIR}/test_2.fastq
kallisto quant -i ${KALLISTO_INDEX} -o ${OUT_DIR}/test_rf ${SEQ_DIR}/test_1.fastq ${SEQ_DIR}/test_2.fastq --rf-stranded
kallisto quant -i ${KALLISTO_INDEX} -o ${OUT_DIR}/test_fr ${SEQ_DIR}/test_1.fastq ${SEQ_DIR}/test_2.fastq --fr-stranded

# infering strandness
paste ${OUT_DIR}/test_fr/abundance.tsv ${OUT_DIR}/test_rf/abundance.tsv ${OUT_DIR}/test_un/abundance.tsv | cut -f1,4,9,14  | awk 'BEGIN{sum1=0;sum2=0;sun3=0}{sum1+=$2;sum2+=$3;sum3+=$4}END{print sum1,sum2,sum3}' > ${OUT_DIR}/test.libtype.txt
cat ${OUT_DIR}/test.libtype.txt | awk '{print $2/$1,$3/$1,$3/$2}' | awk '{if($1<0.3 && $3>3)print "stranded";else if($1>3 && $2>3)print "reverse";else print "unstranded"}' >> ${OUT_DIR}/test.libtype.txt

```


## Quantification of transcripts

Experiment id - GSE120561
```bash
time kallisto quant --single --fragment-length=49 --sd=1 -b 1000 -i kallisto/index/index_trans -o kallisto/quant/w1 rawFastq/SRR7908186_W1_Worker_Pool_1.fastq.gz
time kallisto quant --single --fragment-length=49 --sd=1 -b 1000 -i kallisto/index/index_trans -o kallisto/quant/w2 rawFastq/SRR7908187_W2_Worker_Pool_2.fastq.gz
time kallisto quant --single --fragment-length=49 --sd=1 -b 1000 -i kallisto/index/index_trans -o kallisto/quant/aw1 rawFastq/SRR7908188_AW1_Active_Pool_1.fastq.gz
time kallisto quant --single --fragment-length=49 --sd=1 -b 1000 -i kallisto/index/index_trans -o kallisto/quant/aw2 rawFastq/SRR7908189_AW2_Active_Pool_2.fastq.gz
time kallisto quant --single --fragment-length=49 --sd=1 -b 1000 -i kallisto/index/index_trans -o kallisto/quant/q1 rawFastq/SRR7908190_Q1_Queen_Pool_1.fastq.gz
time kallisto quant --single --fragment-length=49 --sd=1 -b 1000 -i kallisto/index/index_trans -o kallisto/quant/q2 rawFastq/SRR7908191_Q2_Queen_Pool_2.fastq.gz

# or
for i in $(ls rawFastq/); do time kallisto quant -t 4 -b 1000 -i kallisto/index/index_trans ${i} -o kallisto/quant/${i}; done
```


## Output files:

a) abundance.h5: information of bootstraps, which will be used by Sleuth.   
b) abundance.tsv: quantifications of all trancripts in the file all_transcripts.fasta   
c) run_info.json: informations of pseudo alignments  


## Verifying abundance values for lncRNA
```bash
grep lncRNA kallisto/*/abundance.tsv | head
```
target_id         length  eff_length      est_counts      tpm  
aw1.tsv:lncRNA      1152    1104    3       1.12067  
aw2.tsv:lncRNA      1152    1104    7       2.80562  
q1.tsv:lncRNA       1152    1104    7       2.76799  
q2.tsv:lncRNA       1152    1104    4       1.59035  
w1.tsv:lncRNA       1152    1104    183     57.825  
w2.tsv:lncRNA       1152    1104    153     45.2154  



## References

1 - [Bray, N., Pimentel, H., Melsted, P. et al. Near-optimal probabilistic RNA-seq quantification. Nat Biotechnol 34, 525â€“527 (2016).](https://doi.org/10.1038/nbt.3519)  
2 - [How kallisto works](https://bioinfo.iric.ca/understanding-how-kallisto-works/)  
3 - [about kallisto](https://pachterlab.github.io/kallisto/about)  